### ISSUE TITLE
DNS caching and DNS query statistics.

### ISSUE THREAD

**Issue by MasterArmory** [AdguardTeam/AdGuardHome#1453]
DNS Caching:
This is helpful in speeding up requests. If it's currently implemented, I am not aware of it. 

DNS Query Statistics:
This is a feature form PiHole. On the dashboard, I would like to see a pie chart of where my DNS resolutions are originating.  As a bouns, the average response time from each source would also be good to know. 

Thank you.

**Comment by ameshkov** [comment]
> This is helpful in speeding up requests. If it's currently implemented, I am not aware of it.

It is, there are no stats, though.

> This is a feature form PiHole. On the dashboard, I would like to see a pie chart of where my DNS resolutions are originating. As a bouns, the average response time from each source would also be good to know.

Could you please explain what exact stats on this pie chart would you like to see?

**Comment by MasterArmory** [comment]
> Could you please explain what exact stats on this pie chart would you like to see?

Yes, I would like to see the total for responses for each DNS Server used and the cache. 
For example: If I have 200 DNS lookups.  
I would like to see a chart showing:
- 85 came from the cache
- 62 came from 9.9.9.9
- 50 came from 1.1.1.1
- 3 came from 8.8.8.8

Extra credit is if in the pie chart or graph you also showed the average response time for each DNS Server. 
- 85 came from the cache -Agv. 10ms
- 62 came from 9.9.9.9 -Avg. 100ms
- 50 came from 1.1.1.1 -Avg. 32ms
- 3 came from 8.8.8.8 -Avg. 24ms

**Comment by ariqmuhammadt** [comment]
I think it doesn't have dns caching because i can't find it listed as a feature, it is better if you add stats or status in the query if it cached or not and response time like in the PiHole. Thanks

**Comment by ameshkov** [comment]
Well, it's just so basic that listing it as a feature would be strange.

Currently, you can configure just the DNS cache size:
https://github.com/AdguardTeam/AdGuardHome/wiki/Configuration#configuration-file

**Comment by alexiri** [comment]
Those statistics would be a great addition! It would be nice if we could also see what was in the cache.

**Comment by timkgh** [comment]
+1 for cache stats: hits, misses, eviction because of TTL vs no room in cache, etc. It helps with cache sizing also to see how the cache is being used.

**Comment by timkgh** [comment]
Stats per upstream would be very useful also. There's "Average processing time" but I would like to see numbers per upstream.

**Comment by yuxuan0107** [comment]
我也需要这个功能

**Comment by becks0815** [comment]
Also +1 for more etailed stats. At the moment I am blind concernning relevant settings like TTL data, cache size usage (percentage of cache hits) or different dns servers reaction times, amount of requests sent to the dns servers or even cache usage (percentage of the assigned memory)

**Comment by adrianmace** [comment]
The DNS pie chart would also give insights into how the weighted load balancing algorithm is working. 

My AGH is configured with multiple upstream providers like so, and racing them against each other using `Parallel Requests`.
```
# NextDNS
45.90.28.0
45.90.30.0
# Cloudflare (with enhanced Security)
1.1.1.2
1.0.0.2
# Quad 9 (with ECS)
9.9.9.11
149.112.112.11
# Google
8.8.8.8
8.8.4.4
```

Through the query log I can see that NextDNS is winning most of the time, but I'd like a visualisation with percentage breakdown on the home screen.

**Comment by ely105** [comment]
+1 on this.  It would be nice to have simple cache stats on the dashboard.  Cache hits % is mostly useful to see how effective the cache is for the time period displayed.

**Comment by Kani3D** [comment]
+1 on this

**Comment by peepsnet** [comment]
I want to see this too...
On the dashboard page...


|       Upstream DNS Server        |   Requestes    |       Avg          |
|----------------------------------|----------------|----------------|
|              Cached                        |     12522        |         2ms        |
|           tls://1.1.1.1                     |       9855        |        19ms       |
|    https://1.1.1.1/query             |       5578        |         16ms      |
|                 8.8.8.8                      |         452        |          33ms      |

**Comment by gyCfjSnO** [comment]
It's now past three (3) years since this request was raised 😂😂😂😂

**Comment by schzhn** [comment]
Upstream statistics added in [v0.107.37](https://github.com/AdguardTeam/AdGuardHome/releases/tag/v0.107.37).

**Comment by ainar-g** [comment]
@schzhn, there are still cache statistics to figure out.

**Comment by timkgh** [comment]
@ainar-g It would also be useful to add stats for `parental_block_host` and `safebrowsing_block_host` response times under "Average processing time"

**Comment by ainar-g** [comment]
@timkgh, I'm not entirely sure what you mean, but please make a separate feature request about that.

**Comment by timkgh** [comment]
> @timkgh, I'm not entirely sure what you mean, but please make a separate feature request about that.

If "browsing security" and "parental control" are enabled in "General settings", then `standard-block.dns.adguard.com` and `family-block.dns.adguard.com` are used for these features, via DNS queries. The Dashboard UI should also show DNS stats for these 2 servers, in the "Average processing time" card.
I think it's relevant in this issue as it also relates to DNS stats.

**Comment by zgqq** [comment]
Need to hit cache stats.

**Comment by fvdm** [comment]
I would love to see the cache hit/miss ratio in the API. Especially useful combined with my other LAN stats in Home Assistant.

**Comment by webysther** [comment]
Cache hit rate will be nice, at least overall: "Served from cache %".

**Comment by Plopaplopa** [comment]
i'd love to have cache stats

**Comment by muchqs** [comment]
definitely would like to see cache hit %, as right now I seem to have problem hitting cache. lots entries in logs showing high ping despite "optimistic cache" option on.

**Comment by jonastr** [comment]
Coming from Pi-Hole, this is the one feature I miss. Everything else Adguard does better but c'mon - this is super relevant if one wants to optimize cache size, for example.

As a side note, reqs/sec would be another interesting metric to see the ongoing load.

**Comment by Haravikk** [comment]
Was about to post a request for the same thing.

I'd love to see this implemented as an additional graph option for the top row of the dashboard, with the following behaviour:

- If adult content filtering is disabled, simply replace that graph (it's just wasting space to show a bunch of zeroes).
- If adult content filtering is enabled, combine it with the malware/phishing graph (two lines on the same graph).

**Comment by Sil3ntVip3r** [comment]
I traced the remaining cache-statistics part against current `master`.  The
upstream-response portion has already shipped; cache hits are the missing
piece, and the current code contains most of the raw signal needed to add them.

Current state:

- dnsproxy marks a request's `UpstreamStatistics` entry with `IsCached`;
- the query log consumes that signal and sets its existing `cached` field;
- `internal/stats/unit.go` deliberately skips cached entries while aggregating
  per-upstream counts, but does not increment a separate cache-hit counter;
- the future API contract in `openapi/next.yaml` already reserves
  `dns_cache_hit_rate` and `dns_cache_records` in `GetV1StatsAllResp`; and
- dnsproxy's underlying cache already tracks `Count`, `Size`, `Hit`, and `Miss`,
  but those values are not exposed through `proxy.Proxy`.

That suggests a narrow, testable implementation sequence rather than deriving
the numbers from query-log searches:

1. Persist a per-hour cache-hit count in `internal/stats` alongside the existing
   total-query and result counters.  Increment it when the main query statistic
   is cached.
2. Aggregate `cache_hits / eligible_queries * 100` over the requested stats
   interval.  Define the denominator explicitly (I recommend DNS requests that
   reached dnsproxy's cache/upstream path, excluding locally answered rewrites,
   DHCP answers, and blocked requests).
3. Expose current cache record count through a small read-only dnsproxy API that
   sums the normal and ECS caches under their existing locks.
4. Add the two fields to the active stats API/UI only after confirming whether
   the project wants to advance the `next.yaml` contract or backport them to the
   legacy `/control/stats` response.

Regression coverage should include cache disabled, miss then hit, optimistic
hit, ECS/non-ECS caches, local rewrite and blocked requests excluded from the
denominator, persistence/restart of hourly hit counts, stats reset, and an
empty denominator returning `0` rather than NaN.

The remaining product decision is the API boundary: should this land only in
the versioned `v1` stats work already described by `openapi/next.yaml`, or also
extend the current `/control/stats` response?  Once maintainers choose that,
the backend collection change can stay independent of the dashboard
visualization.
